TP réalisé par :
- Angéline MARC
- Thomas HUGUENEL

In [25]:
import numpy as np 
np.set_printoptions(threshold=10000, suppress = True) 
import pandas as pd 
import warnings 
import matplotlib.pyplot as plt 
warnings.filterwarnings('ignore')

from utilsfunction import *

%load_ext autoreload
%autoreload 2
import importlib, utilsfunction
importlib.reload(utilsfunction)

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


<module 'utilsfunction' from 'c:\\Users\\runhi\\Desktop\\TAA_TP2\\utilsfunction.py'>

In [14]:
data = pd.read_csv('PubMed-multi-label-dataset.csv', sep=",")
data.head(3)

Title  \
0  Expression of p53 and coexistence of HPV in pr...   
1  Vitamin D status in pregnant Indian women acro...   
2  [Identification of a functionally important di...   

                                        abstractText  \
0  Fifty-four paraffin embedded tissue sections f...   
1  The present cross-sectional study was conducte...   
2  The occurrence of individual amino acids and d...   

                                           meshMajor      pmid  \
0  ['DNA Probes, HPV', 'DNA, Viral', 'Female', 'H...   8549602   
1  ['Adult', 'Alkaline Phosphatase', 'Breast Feed...  21736816   
2  ['Amino Acid Sequence', 'Analgesics, Opioid', ...  19060934   

                                              meshid  \
0  [['D13.444.600.223.555', 'D27.505.259.750.600....   
1  [['M01.060.116'], ['D08.811.277.352.650.035'],...   
2  [['G02.111.570.060', 'L01.453.245.667.060'], [...   

                                            meshroot  A  B  C  D  E  F  G  H  \
0  ['Chemicals and Drugs [D]', 'Organisms [B]', '...  0  1  1  1  1  0  0  1   
1  ['Named Groups [M]', 'Chemicals and Drugs [D]'...  0  1  1  1  1  1  1  0   
2  ['Phenomena and Processes [G]', 'Information S...  1  1  0  1  1  0  1  0   

   I  J  L  M  N  Z  
0  0  0  0  0  0  0  
1  1  1  0  1  1  1  
2  0  0  1  0  0  0

## Nettoyage des données

In [19]:
data = prepare_data(data)
data.head(3)


abstractText  A  B  C  D  E  F  G  H  \
0  fiftyfour paraffin embedded tissue sections pa...  0  1  1  1  1  0  0  1   
1  the present crosssectional study conducted det...  0  1  1  1  1  1  1  0   
2  the occurrence individual amino acids dipeptid...  1  1  0  1  1  0  1  0   

   I  J  L  M  N  Z  
0  0  0  0  0  0  0  
1  1  1  0  1  1  1  
2  0  0  1  0  0  0

In [21]:
# Séparation des données en données d'entrainement et de test 50-50
X_train, X_test, y_train, y_test = split_data(data)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(25000,) (25000,) (25000, 14) (25000, 14)


In [22]:
#TF-IDF
X_train_tfidf, X_test_tfidf, vectorizer = tfidf_vectorizer(X_train, X_test)

In [14]:
# Evaluation des modèles sur les données
evaluations = run_models(X_train_tfidf, y_train, X_test_tfidf, y_test)

{'model': 'Classifier Chain + Random Forest', 'F1 Score (micro)': 0.8244803122553863, 'F1 Score (macro)': 0.6462110550075353, 'Zero-One Loss': 0.87004}
{'model': 'Classifier Chain + MLP', 'F1 Score (micro)': 0.8085315751925016, 'F1 Score (macro)': 0.6963527663534999, 'Zero-One Loss': 0.88664}
{'model': 'Classifier Chain + KNN', 'F1 Score (micro)': 0.7724249865623979, 'F1 Score (macro)': 0.6233706323311827, 'Zero-One Loss': 0.91776}
{'model': 'Multi Output Classifier + Random Forest', 'F1 Score (micro)': 0.8244920233311362, 'F1 Score (macro)': 0.6457706634712633, 'Zero-One Loss': 0.87088}
{'model': 'Multi Output Classifier + MLP', 'F1 Score (micro)': 0.8105228386081065, 'F1 Score (macro)': 0.696458492330284, 'Zero-One Loss': 0.88752}
{'model': 'Multi Output Classifier + KNN', 'F1 Score (micro)': 0.7781751990254129, 'F1 Score (macro)': 0.6292545395952757, 'Zero-One Loss': 0.92244}


```python
{'model': 'Classifier Chain + Random Forest', 'F1 Score (micro)': 0.8244803122553863, 'F1 Score (macro)': 0.6462110550075353, 'Zero-One Loss': 0.87004}
{'model': 'Classifier Chain + MLP', 'F1 Score (micro)': 0.8085315751925016, 'F1 Score (macro)': 0.6963527663534999, 'Zero-One Loss': 0.88664}
{'model': 'Classifier Chain + KNN', 'F1 Score (micro)': 0.7724249865623979, 'F1 Score (macro)': 0.6233706323311827, 'Zero-One Loss': 0.91776}
{'model': 'Multi Output Classifier + Random Forest', 'F1 Score (micro)': 0.8244920233311362, 'F1 Score (macro)': 0.6457706634712633, 'Zero-One Loss': 0.87088}
{'model': 'Multi Output Classifier + MLP', 'F1 Score (micro)': 0.8105228386081065, 'F1 Score (macro)': 0.696458492330284, 'Zero-One Loss': 0.88752}
{'model': 'Multi Output Classifier + KNN', 'F1 Score (micro)': 0.7781751990254129, 'F1 Score (macro)': 0.6292545395952757, 'Zero-One Loss': 0.92244}
```

## SVD

In [26]:
X_train_svd, X_test_svd, SVD = svd(X_train_tfidf, X_test_tfidf, 100)

In [27]:
print_top_words(SVD, vectorizer.get_feature_names_out(), 10) # pour afficher les 10 mots les plus pertinents par concept 

Concept #0: patients the cells group study treatment in we cell cancer
Concept #1: patients years group risk age months 95 ci patient care
Concept #2: patients cells cell cancer tumor expression survival treatment therapy tumors
Concept #3: cancer health cells care cell breast expression tumor risk gene
Concept #4: women cells ci 95 cancer risk group levels health age
Concept #5: cells health care treatment cell patient group pain medical mice
Concept #6: patients health gene expression protein genes care group levels dna
Concept #7: cancer group breast treatment tumor patients groups survival prostate health
Concept #8: group expression gene children genes cases groups rats case mice
Concept #9: group cells children groups dna cell the method bone 95
Concept #10: children treatment cases infection acid serum virus liver hiv activity
Concept #11: health group care blood artery coronary pressure protein pulmonary dna
Concept #12: blood insulin glucose serum levels samples plasma method 

In [19]:
# Evaluation des modèles sur les données après SVD
evaluations = run_models(X_train_svd, y_train, X_test_svd, y_test)

{'model': 'Classifier Chain + Random Forest', 'F1 Score (micro)': 0.814969172270081, 'F1 Score (macro)': 0.6293764324843805, 'Zero-One Loss': 0.87724}
{'model': 'Classifier Chain + MLP', 'F1 Score (micro)': 0.7758129960334826, 'F1 Score (macro)': 0.6560806376044893, 'Zero-One Loss': 0.92136}
{'model': 'Classifier Chain + KNN', 'F1 Score (micro)': 0.7749726851330897, 'F1 Score (macro)': 0.6307600888876275, 'Zero-One Loss': 0.9094}
{'model': 'Multi Output Classifier + Random Forest', 'F1 Score (micro)': 0.8141937104210394, 'F1 Score (macro)': 0.6246835529531835, 'Zero-One Loss': 0.88032}
{'model': 'Multi Output Classifier + MLP', 'F1 Score (micro)': 0.7756432330526308, 'F1 Score (macro)': 0.6558157794806434, 'Zero-One Loss': 0.93152}
{'model': 'Multi Output Classifier + KNN', 'F1 Score (micro)': 0.7842723755781146, 'F1 Score (macro)': 0.6473781796318662, 'Zero-One Loss': 0.9106}


{'model': 'Classifier Chain + Random Forest', 'F1 Score (micro)': 0.814969172270081, 'F1 Score (macro)': 0.6293764324843805, 'Zero-One Loss': 0.87724}

{'model': 'Classifier Chain + MLP', 'F1 Score (micro)': 0.7758129960334826, 'F1 Score (macro)': 0.6560806376044893, 'Zero-One Loss': 0.92136}

{'model': 'Classifier Chain + KNN', 'F1 Score (micro)': 0.7749726851330897, 'F1 Score (macro)': 0.6307600888876275, 'Zero-One Loss': 0.9094}

{'model': 'Multi Output Classifier + Random Forest', 'F1 Score (micro)': 0.8141937104210394, 'F1 Score (macro)': 0.6246835529531835, 'Zero-One Loss': 0.88032}

{'model': 'Multi Output Classifier + MLP', 'F1 Score (micro)': 0.7756432330526308, 'F1 Score (macro)': 0.6558157794806434, 'Zero-One Loss': 0.93152}

{'model': 'Multi Output Classifier + KNN', 'F1 Score (micro)': 0.7842723755781146, 'F1 Score (macro)': 0.6473781796318662, 'Zero-One Loss': 0.9106}

## Interprétation
Nous avons un modèle légérement moins performant après l'application de la SVD. Cela peut être dû à la perte d'information lors de la réduction de la dimensionnalité. En effet, les différents F1 scorres ont un peu baissé, mais la différence n'est pas significative. A l'inverse, le zero-one loss a augmenté pour la plupart des modèles, ce qui signifie que le modèle est moins performant.

Cependant, la différence n'est pas significative (un différence de 0.01 dans beaucoup de cas) et la SVD a permis de réduire le temps de calcul d'une vingtaine de minutes.